# Background

Down syndrome (DS), or trisomy 21, is caused by an extra copy of human chromosome 21 (HSA21) and is characterized by a range of anatomical and cellular abnormalities which lead to intellectual disabilities (ID), developmental delays, and an early onset of Alzheimerâ€™s disease (AD)1. DS-associated ID has been attributed to neurodevelopmental deficits like reduced cortical volume, reduced proliferating neural progenitor cells (NPCs), neuronal hypocellularity, delayed and decreased myelination, and reduced gliogenesis2. However, the molecular mechanisms that underscore these deficits are not yet well understood.

Induced pluripotent stem cells (iPSCs) derived from individuals with DS have been used to study the neurodevelopmental changes that contribute to ID, allowing for a longitudinal study of cellular differentiation and maturation. While these 2D cultures provide a useful human genetic system for DS, they do not mimic the complexity and organization of the human brain3. iPSC-derived 3D cultures allow the study of more complex interactions between cells, and provide spatial organization, allowing for a more relevant model of the human brain for DS studies3, however this model still does not fully recapitulate the heterogeneity and complexity of the human brain.

Forebrain assembloids (FAs) are iPSC-derived 3D cultures that are more translationally relevant than previously mentioned models, as they offer the benefits of spatial organization/complexity of traditional 3D cultures, while providing more tissue heterogeneity, making this model a closer representation of the human brain. These assembloids combine dorsal and ventral forebrain-like organoids, culminating to a model that allows for a deeper understanding of neural development as it relates to cell migration and connectivity.

Single-cell sequencing techniques are highly applicable to the characterization of FAs, as they can capture the diversity and heterogeneity of cell types within the culture. This granularity is not achievable with traditional bulk techniques which average expression across many cells, potentially masking the unique expression profiles of small cell populations. Single-cell RNA sequencing (scRNA-seq) allows for transcriptomic characterization of FAs, providing insight on differences in cellular composition between euploid and triploid FAs. Additionally, this information can be used to identify DS specific gene expression signatures, and map developmental trajectories of cells within the FAs. Single-cell ATAC sequencing profiles chromatin accessibility at the single-cell level, providing insights into the regulatory landscape and epigenetic changes in DS FAs. This information can be used to identify altered regulatory elements in specific cell types, and characterize the epigenetic landscape during DS development. In combination, these 2 approaches provide a comprehensive understanding of the transcriptional and epigenetic changes that contribute to DS pathology. This integrated
approach links gene expression patterns with chromatin accessibility, enabling the identification of key regulatory networks and transcription factor binding sites that drive the aberrant neurodevelopment apparent in DS. In this study, the transcriptomic and epigenomic profiles of FAs from DS patients will be analyzed to gain a more robust cellular and molecular understanding of the neurodevelopmental deficits that contribute to ID in DS.
